### PR TITLE
Move platform check into modules

### PR DESCRIPTION
**Pull request by rycee**
Before, loading a module would be guarded by an optional platform condition. This made it possible to avoid loading and evaluating a module if it did not support the host platform.

Unfortunately, this made it impossible to share a single configuration between GNU/Linux and Darwin hosts, which some wish to do.

This removes the conditional load and instead inserts host platform assertions in the modules that are platform specific.

### CODE REVIEW

File: modules/services/sxhkd.nix [hunk 1]
```diff
@@ -65,6 +65,8 @@ in
   };
 
   config = mkIf cfg.enable {
+assertions = [(lib.hm.assertions.assertPlatform "services.sxhkd" pkgs lib.platforms.linux)];
+
```
Comment (berbiche) [inline modules/services/sxhkd.nix:69]: This is not properly formatted.

File: modules/services/window-managers/xmonad.nix [hunk 2]
```diff
@@ -137,10 +137,23 @@ in {
       }/bin/xmonad-${pkgs.hostPlatform.system}";
 
   in mkIf cfg.enable (mkMerge [
-    { home.packages = [ (lowPrio xmonad) ]; }
+    {
+      assertions = [
+        (hm.assertions.assertPlatform "xsession.windowManager.awesome" pkgs
```
Comment (berbiche) [inline modules/services/window-managers/xmonad.nix:142]:
```suggestion
        (hm.assertions.assertPlatform "xsession.windowManager.xmonad" pkgs
```

**Comment by berbiche** [comment]
I'll test this on my Mac and NixOS as soon as possible.

File: modules/services/window-managers/xmonad.nix (hunk 2 above)
Comment (rycee) [reply modules/services/window-managers/xmonad.nix]: Thanks, fixed now 👍

File: modules/services/sxhkd.nix (hunk 1 above)
Comment (rycee) [reply modules/services/sxhkd.nix]: Indeed, fixed now.

**Comment by league** [comment]
Hi, I'm trying it out. On NixOS Linux, my flake evaluated fine and `nix flake check` passed. It's rebuilding system and user profiles now. (For some reason it decided it needed to rebuild `emacs-gcc` although I don't *think* my pinned `nix-community/emacs-overlay` input changed.)

On Mac though, something went wrong. Could easily be something strange I'm doing, or just the fact that this is substantially newer than the home-manager I had pinned. (I'm getting the warning about Home Manager version 21.11 and Nixpkgs version 21.05).

Anyway, my previously-pinned home-manager was 42847469b3f65a363dc52b66be09d0ac4edcc55c and on my very old Mac, `nix flake check` runs fine:

```
warning: unknown flake output 'lib'                                                             
trace: lib.crossLists is deprecated, use lib.cartesianProductOfSets instead                     
trace: `mkStrict' is obsolete; use `mkOverride 0' instead.                                      trace: `lib.nixpkgsVersion` is deprecated, use `lib.version` instead!
trace: warning: lib.readPathsFromFile is deprecated, use a list instead                         
trace: Warning: `showVal` is deprecated and will be removed in the next release, please use `tra
ceSeqN`                                                                                         
trace: lib.zip is deprecated, use lib.zipAttrsWith instead                                      
warning: unknown flake output 'hmModules'                                                       
warning: unknown flake output 'hmConfigurations' 
```

But when I update the pin to this PR 4140249ff496effeda8e2a2acc853009c8716b99 it goes wrong:

```
warning: Git tree '/Users/league/u/nix-prime' is dirty            
warning: unknown flake output 'lib'                                                             
trace: warning: league profile: You are using                                                   
                                                                                                
  Home Manager version 21.11 and                                                                
  Nixpkgs version 21.05.
[blah blah... several of these]
error: --- Error --------------------------------------------------------------------------- nix
a 'x86_64-linux' with features {} is required to build '/nix/store/v8chp8gakdi2z7rf5aba060gkkf19
n23-source.drv', but I am a 'x86_64-darwin' with features {benchmark, big-parallel, nixos-test}
(use '--show-trace' to show detailed location information)
```

So it's trying to evaluate something dependent on `x86_64-linux`. I'll attach the [contents of that drv](https://github.com/nix-community/home-manager/files/6792831/v8chp8gakdi2z7rf5aba060gkkf19n23-source.drv.txt). It seems to reference `rycee/nur-expressions` on gitlab and indeed I am using `nur.repos.rycee.firefox-addons`, which shows up in [the output with show-trace turned on](https://github.com/nix-community/home-manager/files/6792862/show-trace.txt)

Again, entirely possible this breaks sometime before this PR... I can investigate further (such as trying it with master) but I'm running out of time I can play with it today. Don't necessarily let this hold you up!

**Comment by league** [comment]
If it matters, here are the firefox extensions I reference:

```nix
{   programs.firefox.extensions = with pkgs.nur.repos.rycee.firefox-addons;
        [
          https-everywhere
          lastpass-password-manager
          temporary-containers
          textern
        ] ++ lib.optionals cfg.plasma-integration [ plasma-integration ];
}
```

Oh hang on, good chance that `plasma-integration` is Linux-only? Actually I think *none* of this firefox config is being **used** by my MacOS profile. It was just hanging out there and `nix flake check` was able to evaluate it before, and now it's not.

**Comment by rycee** [comment]
@league Thanks for testing! Very helpful. Would you mind verifying that you don't get the same error with the commit before this one, 06ee8ec8dfc4554c374052cd7b7083765748af99? I imagine it is likely that this commit introduces the problem, but best to be certain 🙂

In any case, as far as I can tell, plasma-integration does indeed have `platforms = lib.platforms.linux` but from what I can tell the error is not related to that.

The problem seems to be in the drv you linked. For some reason Nix decides to require a `x86_64-linux` system just to download my NUR. Very curious, I cannot imagine what would cause that. If you try commenting out the part of your configuration that references `pkgs.nur`, does it go further in the build?

**Comment by league** [comment]
Interesting… the problem also occurs on 06ee8ec8dfc4554c374052cd7b7083765748af99, so it's not really this PR. Maybe I'll attempt to bisect between 42847469b3f65a363dc52b66be09d0ac4edcc55c and 06ee8ec8dfc4554c374052cd7b7083765748af99 to see when it breaks!

**Comment by league** [comment]
Oh and to answer your other question, if I remove the reference to `pkgs.nur` for the firefox extensions, the problem does go away.

Strangely enough, the offending commit for this issue appears to be 2aeaf65e8f9219c1acdb47bcf278983b3170a344. Its predecessor, f4998f0adccc60a2b463f1892e3eb42b9715a8ee did not exhibit this problem. I don't know what it means yet.

**Comment by rycee** [comment]
@league Thanks for the bisect, very curious. So, from what I can tell 2aeaf65e8f9219c1acdb47bcf278983b3170a344 introduces a dependency on `programs.firefox.extensions` that wasn't there before, which in turn pulls in `pkgs.nur`, which fails because `pkgs` is set up for the `x86_64-linux` platform? I've got to admin that I feel quite puzzled about this 😖 I will have to ponder a bit on that one…

**Comment by league** [comment]
@rycee Yes, I think you characterized it well! A test that triggers the issue independent of my config may help, but I'm swamped at the moment.

**Comment by rycee** [comment]
Since the error @league ran in to doesn't seem to be related to this one I figure it should be OK to merge for now. We'll have to hunt down the error in a separate issue.

**Pull request merged**